src/wh_server_keystore.c

```diff
@@ -0,0 +1,293 @@
+            ctx->nvmMetaCache[i].len);
+    }
+    /* calculate key digest */
+    if (ret == 0)
+        ret = wc_InitSha256_ex(sha, ctx->heap, ctx->devId);
+    if (ret == 0)
+        ret = wc_Sha256Update(sha, out, ctx->nvmMetaCache[i].len);
+    if (ret == 0)
+        ret = wc_Sha256Final(sha, digest);
+    /* validate key */
+    if (XMEMCMP(digest, ctx->nvmMetaCache[i].confDigest, WOLFHSM_DIGEST_STUB)
+        != 0) {
+        ret = BAD_FUNC_ARG;
+    }
+    if (ret == 0) {
+        /* cache key if free slot, will only kick out other commited keys */
+        hsmCacheKey(ctx, &ctx->nvmMetaCache[i], out);
+        /* update out meta with nvm cache contents */
+        XMEMCPY(meta, &ctx->nvmMetaCache[i], sizeof(NvmMetaData));
+    }
+#ifdef WOLFHSM_SHE_EXTENSION
+    /* use empty string if we couldn't find the master ecu key */
+    if (ret != 0 && searchId == WOLFHSM_SHE_MASTER_ECU_KEY_ID) {
+        XMEMSET(out, 0, WOLFHSM_SHE_KEY_SZ);
+        meta->len = WOLFHSM_SHE_KEY_SZ;
+        meta->id = searchId;
+        ret = 0;
+    }
+#endif
+    return ret;
+#endif
+}
+
+int hsmEvictKey(whServerContext* server, uint16_t keyId)
+{
+    int ret = 0;
+    int i;
+    /* find key */
+    for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+        /* mark key as erased */
+        if (server->cache[i].meta->id == keyId) {
+            server->cache[i].meta->id = WOLFHSM_ID_ERASED;
+            break;
+        }
+    }
+    /* if the key wasn't found return an error */
+    if (i >= WOLFHSM_NUM_RAMKEYS)
+        ret = BAD_FUNC_ARG;
+    return ret;
+}
+
+int _wh_Server_HandleKeyRequest(whServerContext* server,
+        uint16_t magic, uint16_t action, uint16_t seq,
+        uint8_t* data, uint16_t* size)
+{
+    int ret = 0;
+    uint8_t* in;
+    uint8_t* out;
+    whPacket* packet = (whPacket*)data;
+    whNvmMetadata meta[1] = {0};
+    switch (action)
+    {
+    case WH_KEY_CACHE:
+        /* in is after fixed size fields */
+        in = (uint8_t*)(&packet->keyCacheReq + 1);
+        /* set the metadata fields */
+        meta->id = packet->keyCacheReq.id;
+        meta->flags = packet->keyCacheReq.flags;
+        meta->len = packet->keyCacheReq.len;
+        XMEMCPY(meta->label, packet->keyCacheReq.label, WOLFHSM_NVM_LABEL_LEN);
+        /* get a new id if one wasn't provided */
+        if (meta->id == WOLFHSM_ID_ERASED) {
+            ret =  hsmGetUniqueId(server);
+            if (ret > 0) {
+                meta->id = ret;
+                ret = 0;
+            }
+        }
+        /* write the key */
+        if (ret == 0)
+            ret = hsmCacheKey(server, meta, in);
+        if (ret == 0) {
+            packet->keyCacheRes.id = meta->id;
+            *size = WOLFHSM_PACKET_STUB_SIZE + sizeof(packet->keyCacheRes);
+        }
+        break;
+    case WH_KEY_EVICT:
+        ret = hsmEvictKey(server, packet->keyEvictReq.id);
+        if (ret == 0) {
+            packet->keyEvictRes.ok = 0;
```

Comment:
Looks like `ok` will always be zero, even on error, unless the client explicitly initializes it to a non-zero value (which `wh_Client_KeyEvictRequest` does not). We should instead report a generic rc value that is explicitly set in both success and failure cases. Same goes for the `ok` fields for the rest of this switch statement.

Reply:
I do, if an error occurs it is set on packet->rc, the ok is just a placeholder so the packet has it's body in case something else needs to be added, could probably remove it